src/porepy/numerics/ad/grid_operators.py

```diff
@@ -250,251 +268,417 @@ def __init__(
+        # whether the interfaces are conforming or not, we need to store different
+        # projections.
+        if self._is_conforming_primary:
+            # Mapping for intensive and extensive quantities are the same, we can get
+            # away with storing only one set of projections.
+            self._mortar_to_primary: Optional[SparseArray] = None
+            self._primary_to_mortar: Optional[SparseArray] = None
         else:
-            num_cells_lower_dimension = sum([g.num_cells for g in subdomains]) * dim
-            num_faces_higher_dimension = sum([g.num_faces for g in subdomains]) * dim
-
-            # Projections to and from the grid
-            to_face = sps.csc_matrix((num_faces_higher_dimension, 0))
-            from_face = sps.csr_matrix((0, num_faces_higher_dimension))
-            to_cells = sps.csc_matrix((num_cells_lower_dimension, 0))
-            from_cells = sps.csr_matrix((0, num_cells_lower_dimension))
-
-            # Projections to primary
-            mortar_to_primary_int.append(to_face)
-            mortar_to_primary_avg.append(to_face)
-
-            # Projections from primary
-            primary_to_mortar_int.append(from_face)
-            primary_to_mortar_avg.append(from_face)
-
-            mortar_to_secondary_int.append(to_cells)
-            mortar_to_secondary_avg.append(to_cells)
-
-            secondary_to_mortar_int.append(from_cells)
-            secondary_to_mortar_avg.append(from_cells)
-
-        # Stack mappings from the mortar horizontally.
-        # The projections are wrapped by a pp.ad.SparseArray to be compatible with the
-        # requirements for processing of Ad operators.
-        def bmat(matrices, name):
-            # Create block matrix, convert it to optimized storage format
-            block_matrix = pp.matrix_operations.optimized_compressed_storage(
-                sps.bmat(matrices)
-            )
-            return SparseArray(block_matrix, name=name)
+            # Here we need to store both intensive and extensive projections.
+            self._mortar_to_primary_int: Optional[SparseArray] = None
+            self._mortar_to_primary_avg: Optional[SparseArray] = None
+            self._primary_to_mortar_int: Optional[SparseArray] = None
+            self._primary_to_mortar_avg: Optional[SparseArray] = None
+
+        # The case for secondary grids is analogous to the primary grids.
+        if self._is_conforming_secondary:
+            self._mortar_to_secondary: Optional[SparseArray] = None
+            self._secondary_to_mortar: Optional[SparseArray] = None
+        else:
+            self._mortar_to_secondary_int: Optional[SparseArray] = None
+            self._mortar_to_secondary_avg: Optional[SparseArray] = None
+            self._secondary_to_mortar_int: Optional[SparseArray] = None
+            self._secondary_to_mortar_avg: Optional[SparseArray] = None
+
+    def sign_of_mortar_sides(self) -> SparseArray:
```

Comment:
I did not get where this is still used. I guess outside this class.

Reply:
Yes.